components/tabs/src/sync/bridge.rs

```diff
@@ -0,0 +1,444 @@
+    }
+
+    fn prepare_for_sync(&self, client_data: &str) -> Result<()> {
+        let data: ClientData = serde_json::from_str(client_data)?;
+        self.sync_impl.lock().unwrap().prepare_for_sync(data)
+    }
+
+    fn sync_started(&self) -> Result<()> {
+        // This is a no-op for the Tabs Engine
+        Ok(())
+    }
+
+    fn store_incoming(&self, incoming_envelopes: &[IncomingEnvelope]) -> Result<()> {
+        // Store the incoming payload in memory so we can use it in apply
+        *(self.incoming_payload.lock().unwrap()) = incoming_envelopes.to_vec();
+        Ok(())
+    }
+
+    fn apply(&self) -> Result<ApplyResults> {
+        let incoming = self.incoming_payload.lock().unwrap();
+
+        // turn them into a TabRecord.
+        let mut records = Vec::with_capacity(incoming.len());
+        for inc in &*incoming {
+            // This error handling is a bit unfortunate, but will soon be removed as we
+            // move towards unifying the bridged_engine with a "real" engine.
+            let payload = match inc.payload() {
+                Ok(p) => p,
+                Err(e) => {
+                    log::warn!("Ignoring invalid incoming envelope: {}", e);
+                    continue;
+                }
+            };
+            let record = match TabsRecord::from_payload(payload) {
+                Ok(r) => r,
+                Err(e) => {
+                    log::warn!("Ignoring invalid incoming tab record: {}", e);
+                    continue;
+                }
+            };
+            records.push(record);
+        }
+
+        let mut sync_impl = self.sync_impl.lock().unwrap();
+        let outgoing = sync_impl.apply_incoming(records)?;
+
+        // Turn outgoing back into envelopes - a bit inefficient going via a Payload.
+        let mut outgoing_envelopes = Vec::with_capacity(1);
+        if let Some(outgoing_record) = outgoing {
+            let payload = Payload::from_record(outgoing_record)?;
+            outgoing_envelopes.push(payload.into());
+        }
+        Ok(ApplyResults {
+            envelopes: outgoing_envelopes,
+            num_reconciled: Some(0),
+        })
+    }
+
+    fn set_uploaded(&self, server_modified_millis: i64, ids: &[SyncGuid]) -> Result<()> {
+        self.sync_impl
+            .lock()
+            .unwrap()
+            .sync_finished(ServerTimestamp::from_millis(server_modified_millis), ids)
+    }
+
+    fn sync_finished(&self) -> Result<()> {
+        *(self.incoming_payload.lock().unwrap()) = Vec::default();
+        Ok(())
+    }
+
+    fn reset(&self) -> Result<()> {
+        self.sync_impl
+            .lock()
+            .unwrap()
+            .reset(EngineSyncAssociation::Disconnected)?;
+        Ok(())
+    }
+
+    fn wipe(&self) -> Result<()> {
+        self.sync_impl.lock().unwrap().wipe()?;
+        Ok(())
+    }
+}
+
+// This is for uniffi to expose, and does nothing than delegate back to the trait.
```

Comment:
We could do the types too, but in many ways I'd rather think about exposing them directly from sync15

However, it doesn't exist *just* to convert the types - uniffi expects a struct here, but BridgedEngine is a trait - so avoid this entirely would involve *both* the types described in UniFFI, but also the ability to tell UniFFI that BridgedEngine is actually a trait (which is what we've been talking about on and off and I hope to get back to soon, but probably not in time to have it before we want to merge this).

Please correct me if I'm wrong though, because I quite hate this :)

Reply:
I think that's right, but let me double check my understanding.  The the path forward would be:

  - Expose `IncomingEnvelope` in `sync15.udl` and implement external types on desktop.  We could then rename `BridgedEngineImpl` to `TabsBridgedEngine` and expose that directly without the need for this wrapper struct.
  - Add support for trait objects in UniFFI, expose the `BridgedEngine` trait in `sync15.udl`, then `tabs.udl` doesn't need to spell out the methods, it can just say that `TabsStore.bridged_engine()` returns a `BridgedEngine`.

(Not necessarily saying that this would need to be done in 2 PRs, maybe it would be simpler to do them both in 1).